hidimstat/desparsified_lasso.py

```diff
@@ -1,79 +1,133 @@
+                 c=c)
+            for i in range(n_features))
+
+    results = np.asarray(results)
+    Z = np.stack(results[:, 0], axis=1)
+    omega_diag = np.stack(results[:, 1])
+
+    return Z, omega_diag
+
+
+def _compute_residuals(X, column_index, alpha=None, gram=None, max_iter=5000,
                        tol=1e-3, method='lasso', c=0.01):
-    """Nodewise Lasso"""
+    """Compute the residuals of the regression of a given column of the
+    design matrix against the other columns"""
 
     n_samples, n_features = X.shape
     i = column_index
 
     X_new = np.delete(X, i, axis=1)
     y = np.copy(X[:, i])
 
-    if method == 'lasso' and Gram is None:
-        Gram_loc = np.dot(X_new.T, X_new)
-    elif method == 'lasso' and Gram is not None:
-        Gram_loc = np.delete(np.delete(Gram, i, axis=0), i, axis=1)
+    if method != 'lasso':
+        ValueError("The only regression method available is 'lasso'")
+
+    if gram is None:
+        gram_loc = False
+    else:
+        gram_loc = np.delete(np.delete(gram, i, axis=0), i, axis=1)
 
-    if method == 'lasso' and alpha is None:
+    if alpha is None:
         k = c * (1. / n_samples)
-        alpha = k * np.max(np.abs(np.dot(X_new, y)))
+        alpha = k * np.max(np.abs(np.dot(y.T, X_new)))
 
-    if method == 'lasso':
-        clf_lasso_loc = Lasso(alpha=alpha, precompute=Gram_loc,
-                              max_iter=max_iter, tol=tol)
+    clf_lasso_loc = Lasso(alpha=alpha, precompute=gram_loc, max_iter=max_iter,
+                          tol=tol)
 
-    elif method == 'lasso_cv':
-        clf_lasso_loc = LassoCV(max_iter=max_iter, tol=tol, cv=3)
+    clf_lasso_loc.fit(X_new, y)
+    z = y - clf_lasso_loc.predict(X_new)
 
-    if method in ['lasso', 'lasso_cv']:
-        clf_lasso_loc.fit(X_new, y)
-        z = y - clf_lasso_loc.predict(X_new)
+    omega_diag_i = n_samples * np.sum(z ** 2) / np.dot(y, z) ** 2
 
-    omega_diag = n_samples * np.sum(z ** 2) / np.dot(y, z) ** 2
+    return z, omega_diag_i
 
-    return z, omega_diag
 
+def desparsified_lasso(X, y, normalize=True, dof_ajdustement=False,
+                       confidence=0.95, max_iter=5000, tol=1e-3,
+                       residual_method='lasso', c=0.01, n_jobs=1,
+                       memory=None, verbose=0):
 
-def desparsified_lasso_confint(X, y, confidence=0.95, max_iter=5000,
-                               tol=1e-3, method="lasso", c=0.01, n_jobs=1):
     """Desparsified Lasso with confidence intervals
 
     Parameters
     -----------
-    X : ndarray or scipy.sparse matrix, (n_samples, n_features)
+    X : ndarray, shape (n_samples, n_features)
         Data.
 
-    y : ndarray, shape (n_samples,) or (n_samples, n_targets)
-        Target. Will be cast to X's dtype if necessary.
+    y : ndarray, shape (n_samples,)
+        Target.
+
+    normalize : bool, optional (default=True)
+        If True, the regressors X will be normalized before regression and
+        the target y will be centered.
```

Comment:
normalize is getting removed from sklearn. Can you tell people to use StandardScaler on their own before?

Reply:
I am not using it as an argument for a Scikit Learn object such as Lasso ! You probably have seen it: it determines if we use the StandardScaler object (or not) before doing the inference.

Reply:
then call this standardize (normalize was a terrible name I chose 10 years ago...)

Reply:
also it seems it's necessary as as you center y to avoid fitting an intercept. You have examples where it's not recommended to do this?

Reply:
Ok for 'standardize'.
In very rare case: binary classification approximated by regression model with the two class being not equally represented or when X is supposed to be already scaled. It might be intersting to avoid this pre-treatment.
Three solution: 1/I keep it like this, 2/ I remove the option of not doing it, 3/I create 'center_Y' which would be independent from standardize (that would operate only on X). I prefer 1/ or 2/ on my side, I think that 3/ would a bit too many options...

Reply:
But the preprocessing isn't / cannot be harmful ?

Reply:
For example in fMRI, you can be happy not to rescale X. But yes, it general it is recommended to rescale and when it is not necessary to do it but you still do it, it will not create bug, and results will still be reliable. I can remove the option and force the preprocessing if you prefer ?

Reply:
for me the centering of y is more related to the fit of an intercept. If you center y you implicitely fit an intercept.

Reply:
Yes, I agree. In some case you could want to center `y` and let `X` as it is. But it is true that this understanding of the method is a bit too advanced may be. So are we agreeing on forcing this preprocessing step ?

Reply:
well I don't know. There must be a good mathematical reason we can explain or document. I don't want to say "just try what works on your data" :)

Reply:
For the desparsified lasso assumptions are X is standardized and y is centered.